File: findMedianInDataStream.js
class MedianFinder {
  constructor() {
    this.max = new PriorityQueue((a, b) => b - a);
    this.min = new PriorityQueue((a, b) => a - b);
  }

  /**
   *
   * @param {number} num
   * @return {void}
   */
  addNum(num) {
    if (!this.min.isEmpty() && num > this.min.front()) {
      this.min.enqueue(num);
    } else {
      this.max.enqueue(num);
    }

    if (this.max.size() > this.min.size() + 1) {
      this.min.enqueue(this.max.dequeue());
    }
    if (this.min.size() > this.max.size() + 1) {
      this.max.enqueue(this.min.dequeue());
    }
  }

  /**
   * @return {number}
   */
  findMedian() {
    if (this.min.size() > this.max.size()) {
      return this.min.front();
    }
    if (this.max.size() > this.min.size()) {
      return this.max.front();
    }
    if (this.max.size() === this.min.size()) {
      return this.max.front() + (this.min.front() - this.max.front()) / 2;
    }
  }
}

File: find_median_in_data_stream.py
class MedianFinder:

    def __init__(self):
        self.large = []  # minheap
        self.small = []  # maxheap

    def addNum(self, num: int) -> None:
        if self.large and num > self.large[0]:
            heapq.heappush(self.large, num)
        else:
            heapq.heappush(self.small, (num * -1))

        if len(self.small) > (len(self.large) + 1):
            heapq.heappush(self.large, (heapq.heappop(self.small) * -1))
        if len(self.large) > (len(self.small) + 1):
            heapq.heappush(self.small, (heapq.heappop(self.large) * -1))

    def findMedian(self) -> float:
        if len(self.small) > len(self.large):
            return (self.small[0] * -1)
        if len(self.large) > len(self.small):
            return self.large[0]
        if len(self.small) == len(self.large):
            return (self.small[0] * -1) + ((self.large[0] - (self.small[0] * -1)) / 2)
